lib/THC/THCReduceApplyUtils.cuh

```diff
@@ -64,13 +73,31 @@ struct TensorInfo {
 template <typename IndexType>
 TensorInfo<IndexType>::TensorInfo(THCState* state,
                                   THCudaTensor* t,
-                                  int reduceDim)
-    : data(NULL), dims(0) {
-  int origDims = THCudaTensor_nDimension(state, t);
-  assert(origDims <= MAX_CUTORCH_DIMS);
-  assert(reduceDim < origDims);
-
+                                  int reduceDim) {
   data = THCudaTensor_data(state, t);
+  dims = THCudaTensor_nDimension(state, t);
+  assert(dims <= MAX_CUTORCH_DIMS);
+
+  for (int i = 0; i < dims; ++i) {
+    sizes[i] = THCudaTensor_size(state, t, i);
+    strides[i] = THCudaTensor_stride(state, t, i);
+  }
+
+  assert(reduceDim == -1 || (reduceDim < dims && reduceDim >= 0));
+
+  if (reduceDim != -1) {
+    sizes[reduceDim] = 1;
+  }
+
+  collapseDims();
```

Comment:
I'll change all the callsites in cutorch that use the default behavior to collapse it explicitly, and remove the separate option.